halo2_gadgets/src/sinsemilla/merkle/chip.rs

```diff
@@ -215,93 +221,95 @@ where
+            )?;
 
             (b_1, b_2, b)
         };
 
-        let c = {
-            // `c = bits 5..=254 of `right`
-            let c = right
-                .value()
-                .map(|value| bitrange_subset(value, 5..(pallas::Base::NUM_BITS as usize)));
-            self.witness_message_piece(layouter.namespace(|| "Witness c"), c, 25)?
-        };
+        // c = bits 5..=254 of `right`
+        let c = MessagePiece::from_subpieces(
+            self.clone(),
+            layouter.namespace(|| "Witness c"),
+            [RangeConstrained::bitrange_of(
+                right.value(),
+                5..(pallas::Base::NUM_BITS as usize),
+            )],
+        )?;
 
+        // hash = SinsemillaHash(Q, 𝑙⋆ || left⋆ || right⋆)
+        //
+        // `hash = ⊥` is handled internally to `SinsemillaChip::hash_to_point`: incomplete
+        // addition constraints allows ⊥ to occur, and then during synthesis it detects
+        // these edge cases and raises an error (aborting proof creation).
+        //
+        // Note that MerkleCRH as-defined maps ⊥ to 0. This is for completeness outside
+        // the circuit (so that the ⊥ does not propagate into the type system). The chip
+        // explicitly doesn't map ⊥ to 0; in fact it cannot, as doing so would require
+        // constraints that amount to using complete addition. The rationale for excluding
+        // this map is the same as why Sinsemilla uses incomplete addition: this situation
+        // yields a nontrivial discrete log relation, and by assumption it is hard to find
+        // these.
         let (point, zs) = self.hash_to_point(
             layouter.namespace(|| format!("hash at l = {}", l)),
             Q,
-            vec![a.clone(), b.clone(), c.clone()].into(),
+            vec![a.inner(), b.inner(), c.inner()].into(),
         )?;
+        let hash = Self::extract(&point);
+
+        // `SinsemillaChip::hash_to_point` returns the running sum for each `MessagePiece`.
+        // Grab the outputs we need for the decomposition constraints.
         let z1_a = zs[0][1].clone();
         let z1_b = zs[1][1].clone();
 
         // Check that the pieces have been decomposed properly.
-        /*
-            The pieces and subpieces are arranged in the following configuration:
-            |  A_0  |  A_1  |  A_2  |  A_3  |  A_4  | q_decompose |
-            -------------------------------------------------------
-            |   a   |   b   |   c   |  left | right |      1      |
-            |  z1_a |  z1_b |  b_1  |  b_2  |   l   |             |
-        */
+        //
+        // The pieces and subpieces are arranged in the following configuration:
+        // |  A_0  |  A_1  |  A_2  |  A_3  |  A_4  | q_decompose |
+        // -------------------------------------------------------
+        // |   a   |   b   |   c   |  left | right |      1      |
+        // |  z1_a |  z1_b |  b_1  |  b_2  |   l   |             |
```

Comment:
`q_decompose` should be `0` here to be consistent with the other configuration comment.